### PR TITLE
gemato/openpgp: correctly handle duplicate keys vs unexpected keys

### PR DESCRIPTION
**Pull request by robbat2**
The old code path had a subtle behavior bug: if an expected key appeared twice in data from a WKD URL, it was then removed entirely.

This happened at one point due to a GPG behavior: when using --export, if --keyring is passed twice, with different keyrings, but those keyrings both contain the key being exported (possibly with different signatures), then the export output will have duplicates of PGP packets present in both keyrings (e.g. UID).

To avoid this, defer the removal of unexpected keys until the main import is completed.

Signed-off-by: Robin H. Johnson <robbat2@gentoo.org>

### CODE REVIEW

File: gemato/openpgp.py [hunk 1]
```diff
@@ -589,18 +589,25 @@ def refresh_keys_wkd(self):
...
-                        raise_on_error=OpenPGPKeyRefreshError)
+                if fpr not in keys:
+                    unexpected_keys.add(fpr)
+
+        for fpr in unexpected_keys:
+            # we need to delete unexpected keys
+            logging.info(
+                f'refresh_keys_wkd(): got unexpected key, will remove: {fpr}')
+            exitst, out, err = self._spawn_gpg(
+                [GNUPG, '--batch', '--delete-keys', fpr],
```
Comment (mgorny) [inline gemato/openpgp.py:608]: Hmm, can't you pass all keys here in one invocation?

Comment (robbat2) [reply gemato/openpgp.py:608]: The prior code was deleting them one at a time, and I opted for as few changes as possible. 

If we did pass many keys at once, we'd need to ensure we didn't exceed arg max. Happy to include if you want that change in this PR as well, vs split out to another PR.

Comment (robbat2) [reply gemato/openpgp.py]: implemented, with batches.

Comment (mgorny) [reply gemato/openpgp.py]: I don't think it makes sense to add this level of complexity to account for removing 128+ keys which is unlikely enough to ever happen (especially giving WKD limits).

Comment (robbat2) [reply gemato/openpgp.py]: So good to just drop the batch stuff entirely, and let it call single deletes?

File: gemato/openpgp.py [hunk 2]
```diff
@@ -589,18 +589,29 @@ def refresh_keys_wkd(self):
...
-                    keys.remove(fpr)
-                else:
-                    # we need to delete unexpected keys
-                    exitst, out, err = self._spawn_gpg(
-                        [GNUPG, '--batch', '--delete-keys', fpr],
-                        raise_on_error=OpenPGPKeyRefreshError)
+                if fpr not in keys:
+                    unexpected_keys.add(fpr)
+
+
```
Comment (mgorny) [inline gemato/openpgp.py:603]:
```suggestion

```

File: gemato/openpgp.py [hunk 3]
```diff
@@ -588,23 +588,46 @@ def refresh_keys_wkd(self):
...
-        # we need to explicitly ensure all keys were fetched
+        # Need to explicitly ensure all keys were fetched
+        # However:
+        # - any key MAY appear 0 or more times.
+        # - expected keys SHOULD be present.
+        # - unexpected keys MAY also be present.
+        #
+        # Track updated & unexpected keys to compare vs expected keys.
+        updated_keys = set()
+        unexpected_keys = set()
```
Comment (mgorny) [inline gemato/openpgp.py:599]: At this point, I'd put all keys printed by GPG into a single set, and then use set operations to consider "updated" and "unexpected" keys.

File: gemato/openpgp.py [hunk 4]
```diff
@@ -588,23 +588,46 @@ def refresh_keys_wkd(self):
...
+            # we need to delete unexpected keys
+            logging.info(
+                f'refresh_keys_wkd(): got unexpected key, will remove: '
+                f'{unexpected_keys}')
+            exitst, out, err = self._spawn_gpg(
+                [GNUPG, '--batch', '--delete-keys'] + list(unexpected_keys),
+                raise_on_error=OpenPGPKeyRefreshError)
+
+        # Check if any of the keys were NOT updated
+        not_updated_keys = set(keys).difference(updated_keys)
```
Comment (mgorny) [inline gemato/openpgp.py:626]: `keys` should be a set already at this point.

File: gemato/openpgp.py (hunk 3 above)
Comment (robbat2) [reply gemato/openpgp.py]: done

File: gemato/openpgp.py (hunk 4 above)
Comment (robbat2) [reply gemato/openpgp.py]: cleaned up.

**Review by mgorny — APPROVED** [review]
Thanks. I'll do a few more cosmetic changes and push it.

**Pull request closed without merging**